plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php

```diff
@@ -238,13 +238,30 @@ public function prepare_items() {
 		 *
 		 * @param array $query_args Arguments to be passed to `wc_get_orders()`.
 		 */
-		$orders      = wc_get_orders( (array) apply_filters( 'woocommerce_order_list_table_prepare_items_query_args', $this->order_query_args ) );
+		$order_query_args = (array) apply_filters( 'woocommerce_order_list_table_prepare_items_query_args', $this->order_query_args );
+
+		// We must ensure the 'paginate' argument is set.
+		$order_query_args['paginate'] = true;
+
+		$orders      = wc_get_orders( $order_query_args );
 		$this->items = $orders->orders;
 
+		$max_num_pages = $orders->max_num_pages;
+
+		// Check in case the user has attempted to page beyond the available range of orders.
+		if ( 0 === $max_num_pages && $this->order_query_args['page'] > 1 ) {
+			$count_query_args          = $order_query_args;
+			$count_query_args['page']  = 1;
+			$count_query_args['limit'] = 1;
+			$order_count               = wc_get_orders( $count_query_args );
+			$max_num_pages             = (int) ceil( $order_count->total / $order_query_args['limit'] );
```

Comment:
Can't we just use `$order_count->max_num_pages` instead of calculating `$max_num_pages` here? For that to work, we would have to _not_ override the `limit` arg, though, so that the one in `$order_query_args` is used instead.

I might be missing something, though, so please let me know.

Reply:
Hi @jorgeatorres,

Yes—that would definitely make the code a lot cleaner/shorter (and it's what I tried first) ... but it doesn't work quite as one might expect.

Suppose we have 100 orders and we are fetching pages of 20 items (so, 5 pages in total). Regardless of whether the data-store is COT or CPT, once we are 'out of bounds' (in the following example I am trying to access page 20, which is not available) we aren't going to get back the information we need:

```php
wc_get_orders( [
	'limit'    => 20,
	'page'     => 20,
	'paginate' => true,
] )->max_num_pages;

# => (int) 0
```

In fact, the same thing can be observed if we use a `WP_Query` (and in the case of the CPT data-store that is presumably the origin of this behavior):

```php
( new WP_Query( [
	'post_type'      => 'shop_order',
	'paged'          => 20,
	'posts_per_page' => 20,
	'post_status'    => 'any',
] ) )->max_num_pages;

# => (int) 0
```

So, in summary, we will get back useful pagination data so long as we are within the bounds of the available data. Outside of that, it zeroes out and we either need to workaround it as I did here or use direct queries (I'm trying to be data-store agnostic in my approach, though).

Please do let me know if you can see another path (or if I'm misunderstanding), though.

Reply:
Yes, I observed that same behavior (which can be argued is confusing), but in any case, I think specifically `$order_count` should be within bounds because we're setting `page` to 1 in the args for that specific query. So if we don't set `limit = 1` we should be basically running the same query as the main query _but_ on page 1, which would give us correct information... or am I missing something @barryhughes?

Reply:
Sorry, @jorgeatorres—I misunderstood your question 😓 

Yes, I believe you are correct. It's more a question of trade-offs. If we retain the same limit (let's say it is 20) then we are also going to fetch upto that number of orders and load them into memory, simply to discard them and perform a redirect (via the call to `set_pagination_args()`). 

With this strategy, we fetch only 1 entity and the arithmetic to calculate `$max_num_pages` is surely very fast. So, while we could shave off a couple of lines of code by keeping `limit` (and in some cases, this may be a lot higher than 20—such as if the merchant configures the number of items per page to 100), it comes with a higher cost.

On the flip-side, your suggestion probably makes the code a bit more readable. What do you think?

Reply:
> With this strategy, we fetch only 1 entity and the arithmetic to calculate $max_num_pages is surely very fast.

I wasn't considering performance when I posted my comment, but that makes sense. I still think it's not ideal that we have to do this, and possibly something to add/change in `OrdersTableQuery`, but right now is probably not the best moment to do that 😅.

Reply:
> I still think it's not ideal that we have to do this, and possibly something to add/change in OrdersTableQuery, but right now is probably not the best moment to do that 😅.

Yep, I agree with both parts; though changing `OrdersTableQuery` might mean we also need a corresponding change in `WC_Order_Data_Store_CPT`. I think it would be a reasonable change to introduce, and could definitely simplify things.